### PR TITLE
Delete SubscriberList's without subscribers on GovDelivery

### PR DESCRIPTION
**Pull request by dwhenry**
For each SubscriberList query GovDelivery for the 
count of subscribers and filter out those that have 
a count of 0 so they can be deleted. Any subscriber 
lists that fail to successfully get data from GovDelivery
are logged for future investigation.

### CODE REVIEW

File: spec/services/gov_delivery/client_spec.rb [hunk 1]
```diff
@@ -1,8 +1,5 @@
 require "rails_helper" # These tests use #strip_heredoc which is in ActiveSupport
 
-require "app/services/gov_delivery/client"
-require "app/services/gov_delivery/request_builder"
```
Comment (dwhenry) [inline spec/services/gov_delivery/client_spec.rb]: We don't need to delete them, but they also don't add any value. 

As I was editing the file anyway I deleted them to help clean it up. I would proceed to do this to each file as I edited it when they had similar useless includes.

**Comment by dwhenry** [comment]
@jennyd Fixed the typos and added some tests around handling errors. I am only catching expected errors from the API, any unexpected error (i.e. 500 response) will still cause the process to explode, but I think this is the right thing to do.

File: spec/services/gov_delivery/client_spec.rb (hunk 1 above)
Comment (jennyd) [inline spec/services/gov_delivery/client_spec.rb]: In that case could you move them into a separate commit to make it clear that it's unrelated to this change?

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      def call
+        group_size = (SubscriberList.count.to_f / thread_count).ceil
+        subscriber_list_sets = SubscriberList.all.each_slice(group_size)
+
+        results = []
+        semaphore = Mutex.new
+        threads = subscriber_list_sets.map do |subscriber_list_set|
+          Thread.new do
+            list_with_count = add_subscribers_count(subscriber_list_set)
+            semaphore.synchronize { results += list_with_count }
```
Comment (jennyd) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: The plurals have got confused here :)

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    end
+
+  private
+
+    def delete_topic(subscriber_list)
+      topic = client.fetch_topic(subscriber_list.gov_delivery_id)
+
+      if topic && topic['subscribers_count'] == 0
+        puts "Deleting: #{subscriber_list.gov_delivery_id}"
+        delete_topic_with_error_catching(subscriber_list.gov_delivery_id)
```
Comment (jennyd) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: We're actually only deleting from GovDelivery here, and not from the local database.

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    context 'when the topic does not exist' do
+      let(:fetch_topics_response) { nil }
+
+      it 'is not included in the list' do
+        expect(subject.with_zero_subscribers).not_to include(subscriber_list)
+      end
+    end
+
+    context '#when the GovDelivery create endpoint has an error' do
```
Comment (jennyd) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: I think you mean `fetch` rather than `create` here.

File: spec/services/gov_delivery/client_spec.rb (hunk 1 above)
Comment (dwhenry) [inline spec/services/gov_delivery/client_spec.rb]: deleted.. will remove in a separate PR once these are all merged.

File: lib/data_hygiene/delete_topics_without_subscribers.rb (hunk 2 above)
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: replaced with `results_for_thread` as that is clearer on it usage

File: spec/services/gov_delivery/client_spec.rb [hunk 5]
```diff
@@ -94,6 +91,81 @@
...
+          'subscribers_count' => 0,
+          'wireless_enabled' => false,
+          'pagewatch_type' => nil,
+          'pages' => [],
+          'visibility' => 'Unlisted'
+        )
+      end
+    end
+
+    context 'when topic dose not exist' do
```
Comment (dwhenry) [inline spec/services/gov_delivery/client_spec.rb]: fixed

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    end
+
+    def with_zero_subscribers
+      @with_zero_subscribers ||= subscriber_lists
+        .select { |c, s| c&.zero? }
+        .map(&:last)
+    end
+
+    def with_gov_delivery_error
+      @sl_with_gov_delivery_error ||= subscriber_lists
```
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: done

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        .select { |c, s| c.nil? }
+        .map(&:last)
+    end
+
+    private
```
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: done

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    def delete_topic(subscriber_list)
+      topic = client.fetch_topic(subscriber_list.gov_delivery_id)
+      [topic && topic['subscribers_count'], subscriber_list]
```
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: deleted

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+      if topic && topic['subscribers_count'] == 0
+        puts "Deleting: #{subscriber_list.gov_delivery_id}"
+        client.delete_topic(subscriber_list.gov_delivery_id)
+      elsif topic.nil?
+        puts "Skipping #{subscriber_list.gov_delivery_id} as it does not appears on GovDelivery"
```
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: fixed

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        threads.each(&:join)
+        puts ''
+        results
+      end
+
+      def add_subscribers_count(subscriber_lists)
+        subscriber_lists.map do |subscriber_list|
+          print '.'
+          topic = client.fetch_topic(subscriber_list.gov_delivery_id)
+          [topic && topic['subscribers_count'], subscriber_list]
```
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: Extract `fetch_topic` to its own method with error catching for know GovDelivery error codes

File: lib/data_hygiene/delete_topics_without_subscribers.rb [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        end
+      end
+
+      def thread_count
+        ENV.fetch('GOV_DELIEVRY_REQUEST_THREAD_COUNT', 20).to_i
```
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: Removed and replaced with constant

File: lib/tasks/data_hygiene.rake [hunk 12]
```diff
@@ -1,3 +1,11 @@
+desc "Delete subscriber lists with subscribers"
```
Comment (dwhenry) [inline lib/tasks/data_hygiene.rake]: fixed

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+      it 'is not included in the list' do
+        expect(subject.with_zero_subscribers).not_to include(subscriber_list)
+      end
+    end
+  end
+
+  describe '#with_gov_delivery_error' do
+    let!(:subscriber_list) { create(:subscriber_list) }
+    let(:client) { double('GovDelivery::Client') }
```
Comment (dwhenry) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: correct, this was missed during refactoring

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    context 'when the topic has subscribers' do
+      let(:fetch_topics_response) { { 'subscribers_count' => 2 } }
+
+      it 'is not included in the list' do
+        expect(subject.with_gov_delivery_error).not_to include(subscriber_list)
+      end
+    end
+
+    context 'when the topic does not exist' do
+      let(:fetch_topics_response) { nil }
```
Comment (dwhenry) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: As above. Have added error catching around `fetch_topic` and `delete_topic` methods and appropriate testing

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+      it 'is included in the list' do
+        expect(subject.with_gov_delivery_error).to include(subscriber_list)
+      end
+    end
+  end
+
+  describe '#call' do
+    let(:expected_log_output) do
```
Comment (dwhenry) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: moved

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      allow(client).to receive(:delete_topic)
+    end
+
+    context 'when the user enters the delete command' do
+      before do
+        input.puts('delete')
+        input.rewind
+      end
+
+      it 'deletes an subscriber lists with 0 sunscribers' do
```
Comment (dwhenry) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: fixed

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+      end
+    end
+
+    context 'when the user does not enter the delete command' do
+      before do
+        input.puts('skip')
+        input.rewind
+      end
+
+      it 'does not delete an subscriber lists' do
```
Comment (dwhenry) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: fixed

File: lib/data_hygiene/delete_topics_without_subscribers.rb (hunk 3 above)
Comment (dwhenry) [inline lib/data_hygiene/delete_topics_without_subscribers.rb]: umm.. yes.. add code and tests for DB deletion as well

File: spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb (hunk 4 above)
Comment (dwhenry) [inline spec/lib/data_hygiene/delete_topics_without_subscribers_spec.rb]: fixed

**Comment by jennyd** [comment]
I've tested this on Integration and the numbers were right afterwards 👍
